modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java

```diff
@@ -0,0 +1,381 @@
+
+    /**
+     * The Destination Type of the messages in the storage queue.
+     */
+    private DestinationType destinationType;
+
+    /**
+     * Reference to {@link MessageFlusher} to deliver messages
+     */
+    private MessageFlusher messageFlusher;
+
+    /**
+     * Reference to slot coordinator to retrieve slots
+     */
+    private SlotCoordinator slotCoordinator;
+
+    /**
+     * Mapping for slot id to slot
+     */
+    private Map<String, Slot> slotTrackerMap;
+
+    MessageDeliveryTask(String destination,
+                        ProtocolType protocolType,
+                        String storageQueueName,
+                        DestinationType destinationType,
+                        SlotCoordinator slotCoordinator,
+                        MessageFlusher messageFlusher) {
+
+        this.destinationName = destination;
+        this.destinationType = destinationType;
+        this.protocolType = protocolType;
+        this.storageQueueName = storageQueueName;
+        this.slotCoordinator = slotCoordinator;
+        this.messageFlusher = messageFlusher;
+        slotTrackerMap = new HashMap<>();
+
+    }
+
+    /**
+     * Slot delivery task
+     * {@inheritDoc}
+     */
+    @Override
+    public TaskHint execute() throws Exception {
+
+        TaskHint taskHint = TaskHint.ACTIVE;
+        MessageDeliveryInfo messageDeliveryInfo =
+                messageFlusher.getMessageDeliveryInfo(destinationName, protocolType, destinationType);
+
+        // Check in memory buffer in MessageFlusher has room
+        if (messageDeliveryInfo.messageBufferHasRoom()) {
+
+            // Get a slot from coordinator.
+            Slot currentSlot = requestSlot(storageQueueName);
+            currentSlot.setDestinationOfMessagesInSlot(destinationName);
+
+            // If the slot is empty
+            if (0 == currentSlot.getEndMessageId()) {
+
+                // If the message buffer in MessageFlusher is not empty send those messages
+                if (log.isDebugEnabled()) {
+                    log.debug("Received an empty slot from slot manager");
+                }
+                boolean sentFromMessageBuffer =
+                        messageFlusher.sendMessagesInBuffer(messageDeliveryInfo, storageQueueName);
+                if (!sentFromMessageBuffer) {
+                    taskHint = TaskHint.IDLE; // Didn't do productive work
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Received slot for storage queue " + storageQueueName + " is: " +
+                                      currentSlot.getStartMessageId() + " - " + currentSlot.getEndMessageId() +
+                                      "Thread Id:" + Thread.currentThread().getId());
+                }
+                List<DeliverableAndesMetadata> messagesRead = getMetaDataListBySlot(storageQueueName, currentSlot);
+
+                if (messagesRead != null && !messagesRead.isEmpty()) {
```

Comment:
Shall we change the if condition to something like below. 
if ((messagesRead != null) && (!messagesRead.isEmpty()))